hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/CatalogReplicaReplicationEndpoint.java

```diff
@@ -0,0 +1,493 @@
+      try {
+        if ((entries.size() == 1) && entries.get(0).getEdit().isMetaEdit()) {
+          try {
+            // When it is a COMMIT_FLUSH event, it can remove all edits in the queue whose
+            // seq# is less flushSeq#. Those events include flush events and
+            // OPEN_REGION/CLOSE_REGION events, as they work like flush events.
+            WALProtos.FlushDescriptor flushDesc =
+              WALEdit.getCommitFlushDescriptor(entries.get(0).getEdit().getCells().get(0));
+            if (flushDesc != null) {
+              cleanQueue(flushDesc.getFlushSequenceNumber());
+            }
+          } catch (IOException ioe) {
+            LOG.warn("Failed to parse {}", entries.get(0).getEdit().getCells().get(0));
+          }
+        }
+        queue.put(entries);
+      } catch (InterruptedException e) {
+
+      }
+    }
+
+    private CompletableFuture<Long> replicate(List<Entry> entries) {
+      CompletableFuture<Long> future = new CompletableFuture<>();
+      FutureUtils.addListener(conn
+        .replay(regionInfo.getTable(), regionInfo.getEncodedNameAsBytes(), entries,
+          regionInfo.getReplicaId(), 1, operationTimeoutNs, location), (r, e) -> {
+        if (e != null) {
+          LOG.warn("Failed to replicate to {}", regionInfo);
+          LOG.warn("Exception: " + e);
+          future.completeExceptionally(e);
+        } else {
+          future.complete(r);
+        }
+      });
+
+      return future;
+    }
+
+    @Override public final void run() {
+      boolean reload = false;
+      boolean sleep = false;
+      List<Entry> entries = null;
+      while (isRunning()) { // we only loop back here if something fatal happened to our stream
+        if (location == null) {
+          // If location is null, keep searching for it. This applies to replica region server
+          // crashes, and waiting to be assigned. reload flag will be set accordingly in case of
+          // error.
+          try {
+            if (sleep) {
+              Thread.sleep(100);
+            }
+            location =
+              locator.getRegionLocation(regionInfo.getStartKey(), regionInfo.getReplicaId(), reload)
+                .get();
+            sleep = false;
+          } catch (ExecutionException ee) {
+            LOG.warn("Error getting location for meta region " + regionInfo + ", " + ee);
+            reload = true;
+            sleep = true;
+            continue;
+          } catch (InterruptedException ie) {
+            LOG.warn("Interrupted while getting location for meta region " +
+              regionInfo + ", " + ie);
+            Thread.currentThread().interrupt();
+          }
+        }
+        try {
+          if (entries == null) {
+            entries = queue.poll(20000, TimeUnit.MILLISECONDS);
+          }
+          if (entries == null) {
+            continue;
+          }
+
+          CompletableFuture<Long> future = replicate(entries);
+
+          try {
+            future.get();
```

Comment:
No timeout?

Reply:
timeout is not needed, as inside replicate(), it already has timeout.